add includedURLs option

wolffberg opened the pull request:
This PR adds an option to include URLs for authentication instead of excluding URLs.

The changes has been tested in a live Traefik instance but I had a few issues I think should be addressed before it can be considered ready for merge.

1. I was unable to run `go test` successfully. I faced a lot of errors and think I might be missing something to run it successfully. Would you be able to test from your end @lukaszraczylo?
2. I wanted to make sure that the plugin either threw an error and didn't authenticate or authenticated everything if both `ExcludedURLs` and `IncludedURLs` were set. I added a check in [settings.go](https://github.com/lukaszraczylo/traefikoidc/compare/main...wolffberg:traefikoidc:feature/included-urls#diff-f994d989d81e7a913e5e8df7c22011e9b2de3e67679fb07d99f88bbc8331a933R214) but it seems like [`(c *Config) Validate()`](https://github.com/lukaszraczylo/traefikoidc/compare/main...wolffberg:traefikoidc:feature/included-urls#diff-f994d989d81e7a913e5e8df7c22011e9b2de3e67679fb07d99f88bbc8331a933L169) is only run during tests and not when the middleware is loaded. An alternative would be to add it in [main.go](https://github.com/lukaszraczylo/traefikoidc/compare/main...wolffberg:traefikoidc:feature/included-urls#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R638) but it seems exessive having to check for every request. Do you have an opinion on how that could best be achieved?

wolffberg commented:
That was fast 🚀

I think it's a good idea to limit the use to use one or the other. Having the option to use both include and exclude will create more confusion and be prone to errors.

I'm not sure which one of these options is better but I think one of them should be implemented.

If both `ExcludedURLs` and `IncludedURLs` are set, Traefik should log an error message and:
- Enforce authentication on all paths (more secure).
- Bypass authentication on all paths (less prone to downtime).
- Fail all requests (more secure and more prone to downtime but also more noticable).

I also tried the tests again from inside `golang:latest` using Docker and now I get better results (although a lot of failing tests after my changes). I will take a look again when we're locked in a way to address above 👷‍♂️

lukaszraczylo commented:
Ahh - by the way - the timeouts tests can take a moment so don't worry about it too much! :)

wolffberg commented:
I fixed the tests in the latest commit but I'm not sure if this is the way you want to go.

I needed multiple `TraefikOidc` instances to test the different combinations of `ExcludedURLs` and `IncludedURLs` so I made `tOidc` into a list and made it so test cases can select the needed instance.

An alternative way could be to overwrite settings from the common `TraefikOidc` instance.

I also added option 3 from my previous comment, failing all requests if `ExcludedURLs` and `IncludedURLs` is set.

Let me know how you feel.

wolffberg commented:
Hi @lukaszraczylo,

Do you have an update on this - Anything I can do to speed things up? 🙏

wolffberg commented:
Closing this PR since we went with a workaround.
